Utilities/Mergers/src/MergerAlgorithm.cxx

```diff
@@ -0,0 +1,109 @@
+{
+  if (target == nullptr) {
+    throw std::runtime_error("Merging target is nullptr");
+  }
+  if (other == nullptr) {
+    throw std::runtime_error("Object to be merged in is nullptr");
+  }
+  if (other == target) {
+    throw std::runtime_error("Merging target and the other object point to the same address");
+  }
+  // fixme: should we check if names match?
+
+  // We expect that both objects follow the same structure, but we allow to add missing objects to TCollections.
+
+  if (auto custom = dynamic_cast<MergeInterface*>(target)) {
+
+    custom->merge(dynamic_cast<MergeInterface* const>(other));
+
+  } else if (auto targetCollection = dynamic_cast<TCollection*>(target)) {
+
+    // We assume that the other has the same structure
+    auto otherCollection = reinterpret_cast<TCollection*>(other);
+
+    auto otherIterator = otherCollection->MakeIterator();
+    while (auto otherObject = otherIterator->Next()) {
+      TObject* targetObject = targetCollection->FindObject(otherObject->GetName());
+      if (targetObject) {
+        // That might be another collection or a concrete object to be merged, we walk on the collection recursively.
+        merge(targetObject, otherObject);
+      } else {
+        // We prefer to clone instead of passing the pointer in order to simplify deleting the `other`.
+        targetCollection->Add(otherObject->Clone());
+      }
+    }
+
+  } else {
+    Long64_t errorCode = 0;
+    TObjArray otherCollection;
+    otherCollection.SetOwner(false);
+    otherCollection.Add(other);
+
+    const char* className = target->ClassName();
+    if (strncmp(className, "TH1", 3) == 0) {
+      errorCode = reinterpret_cast<TH1*>(target)->Merge(&otherCollection);
+    } else if (strncmp(className, "TH2", 3) == 0) {
+      errorCode = reinterpret_cast<TH2*>(target)->Merge(&otherCollection);
+    } else if (strncmp(className, "TH3", 3) == 0) {
+      errorCode = reinterpret_cast<TH3*>(target)->Merge(&otherCollection);
+    } else if (strncmp(className, "THn", 3) == 0) {
+      errorCode = reinterpret_cast<THn*>(target)->Merge(&otherCollection);
+    } else if (strncmp(className, "THnSparse", 8) == 0) {
+      errorCode = reinterpret_cast<THnSparse*>(target)->Merge(&otherCollection);
+    } else if (strcmp(className, "TTree") == 0) {
+      errorCode = reinterpret_cast<TTree*>(target)->Merge(&otherCollection);
+    } else {
+      throw std::runtime_error("Object with type '" + std::string(className) + "' is not one of mergeable type.");
+    }
+    if (errorCode == -1) {
+      throw std::runtime_error("Merging object of type '" + std::string(target->ClassName()) + "' failed.");
+    }
+  }
+}
+
+void deleteTCollections(TObject* obj)
+{
+  // this is not probably the optimal approach, but it should be ok for now
+  if (auto c = dynamic_cast<TCollection*>(obj)) {
+    c->SetOwner(false);
+    auto iter = c->MakeIterator();
+    while (auto element = iter->Next()) {
+      deleteTCollections(element);
+    }
```

Comment:
Naive question: don't you need to delete the TCollection objects themselves ? here we delete only their elements.

Reply:
Hm, indeed, I forgot about that.